File: src/0208-Implement-Trie-(Prefix-Tree)/0208.cpp
static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();

class Node 
{
public:
    Node()  : isWord(false){}
    ~Node()
    {
        for (const auto& item : next) delete item.second; 
    }

    bool isWord;
    unordered_map<char, Node*> next;
};

class Trie 
{
public:
    /** Initialize your data structure here. */
    Trie()  : _root(new Node()) {}
    ~Trie() { delete _root; }
    
    /** Inserts a word into the trie. */
    void insert(string word) 
    {
        auto cur = _root;
        for (auto c : word)
        {
            if (!(cur->next).count(c)) // or if (cur->next[c] == nullptr)
                cur->next[c] = new Node();   
            cur = cur->next[c];
        }
        if (!cur->isWord) cur->isWord = true;
    }
    
    /** Returns if the word is in the trie. */
    bool search(string word) 
    {
        auto cur = _search(word);
        return cur != nullptr and cur->isWord;
    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    bool startsWith(string prefix) 
    {
        return _search(prefix) != nullptr;
    }
private:
    Node* _root;
    
    Node* _search(string word)
    {
        auto cur = _root;
        for (auto c : word)
        {
            if (!cur->next.count(c)) return nullptr;
            cur = cur->next[c];
        }
        return cur;
    }
};

File: src/0208-Implement-Trie-(Prefix-Tree)/0208.js
/**
 * Initialize your data structure here.
 */
var Trie = function() {
    this.next = {};
    this.isWord = false;
};

/**
 * Inserts a word into the trie. 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    let cur = this;
    for (let c of word) {
        if (!cur.next[c]) {
            cur.next[c] = new Trie();
        }
        cur = cur.next[c];
    }
    if (!cur.isWord) cur.isWord = true;
};


Trie.prototype.find = function(word) {
    let cur = this;
    for (let c of word) {
        if (!cur.next[c]) return null;
        cur = cur.next[c];
    }
    return cur;
};

/**
 * Returns if the word is in the trie. 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    let cur = this.find(word);
    if (cur != null && cur.isWord) return true;
    return false;
};

/**
 * Returns if there is any word in the trie that starts with the given prefix. 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    if (this.find(prefix) != null) return true;
    return false;
};